### PR TITLE
[PULL REQUEST] Updates to tagged O3 simulaton for efficiency and correctness - Closes #1109

### LINKED ISSUE #1109: [BUG/ISSUE] Tagged ozone simulation needs to be updated
### What institution are you from?
Peking University

### Description of the problem
(This problem is related to #973, #821, and #965, especially #973, which has not been solved yet).
 
First, I got the same error as #821, then I realized that the unit of Prod/Loss from fullchem simulation is not the same as old version's (say, for v11-02, the unit of Prod_Ox is kg/m3/s, the unit of Loss_Ox is 1/m3/s, see [here](http://geoschemdata.wustl.edu/ExtData/HEMCO/TAGGED_O3/v2014-09/)), while the code of tagged ozone simulation (`GeosCore/tagged_o3_mod.F90`) is the same.

The unit of Prod/Loss now is said to be kg/s, but indeed, it is still molecules/cm3/s. see #965.

I did some modification to convert the unit in the code, see below, but the result os tagged ozone simulation seems inaccurate. The surface O3 concentration is below 20 ppbv, globally.

### Description of troubleshooting performed
So I t modified the `tagged_o3_mod.F90` below to convert the unit.

In line 244, I convert Prod from [molec/cm3/s] to [kg/m3/s], then to get [kg]
```diff
- ! P24H is in kg/m3 per emission time step (ckeller, 9/17/2014).
-    PPROD = P24H(I,J,L) * BOXVL * ( GET_TS_CHEM()/TS_EMIS )
+ ! P24H is in molec/cm3 per emission time step
+    PPROD = P24H(I,J,L) / 6.022e23_fp * 1.e3_fp * 4.8e+1_fp &
+                       * BOXVL * ( GET_TS_CHEM()/TS_EMIS )
```
 In line 527, I also convert Loss from [molec/cm3/s] to [kg/m3/s], then to get [kg], Spc is said to be in [kg]
```diff
-     ! L(O3) is now in [1/m3] (ckeller, 9/17/2014)
-          IF ( State_Met%InTroposphere(I,J,L) ) THEN
-             LL = Spc(I,J,L,N) * L24H(I,J,L) * BOXVL * DT
+    ! L24H is in [molec/cm3/s]
+          IF ( State_Met%InTroposphere(I,J,L) ) THEN
+            LL = L24H(I,J,L) / 6.022e23_fp * 1.e3_fp * 4.8e+1_fp &
+                              * BOXVL * DT
```
The final calculation is based on line 647
`Spc(I,J,L,N) = Spc(I,J,L,N) + PP(I,J,L,N) - LL`
where `PP` is `PROD` above.

I think one main concern is, what's the mass of Ox? In old version, we have Loss_Ox in [1/m3/s] from fullchem simulation, then we multiply Ox concentration in the tagged simulation. However, for now, if we already have Loss_Ox in [molec/cm3/s] or [kg/m3/s] from fullchem simulation, is it ok to directly use it to calculate `Spc(I,J,L,N) = Spc(I,J,L,N) + PP(I,J,L,N) - LL` in tagged simulation?

Another question bothered me is, In #973, @map06 considered all Ox family (O3 + NO2 + 2NO3 + PAN + PPN + ...), does it have influence here?

Given all of the above, I think both the code and wiki of tagged ozone needs to be updated.

Thank you all in advance!

### GEOS-Chem version
V13.2.1

### PR DESCRIPTION
**Pull request by yantosca**
### Overview
Several issues (see #1127, #1109, #973, #828, #569) have been raised by users of the Tagged O3 simulation.  This simulation used to read P(Ox) in kg/cm3/s and L(Ox) in 1/cm3/s from bpchformat files archived from full-chemistry simulations.  The migration from bpch to netCDF diagnostics broke the tagged O3 simulation in GEOS-Chem.

Furthermore, we now use the [ProdLoss History diagnostic collection](http://wiki.seas.harvard.edu/geos-chem/index.php/History_collections_for_chemistry_and_photolysis#The_ProdLoss_Collection) to archive P(Ox) and L(Ox) from the full-chemistry simulations.  These are now archived in units of molec/cm3/s, which are the units that the KPP chemistry solver uses.  This change of diagnostics also meant that the current Tagged O3 simulation was reading in P(Ox) and L(Ox) in the wrong units.

Also, the code structure of the Tagged O3 simulation was confusing and potentially inefficient, as a 3-D parallel loop was being started for each of the 13 tagged species.
 
### Solution
These issues have been addresed in commits 855b305f02df12a8f9f62fbc9c1d5ee3c758fc48. 3bae7571a8c229c0a25144959077fd47a6acb2c4, and 3bae7571a8c229c0a25144959077fd47a6acb2c4.  We have made sure that the unit conversion is correct.  We have also refactored the code for clarity, and now use a single parallel loop that has been optimized for efficiency (with `!$OMP COLLAPSE` and `!$OMP SCHEDULE( DYNAMIC, 4)`.

### Validation
Here is a difference-plot of global O3 from a 1-day Tagged O3 simulation compared to a 1-day fullchem simulation.  24-h averaged P(O3) and L(O3) from the fullchem simulation were read into the Tagged O3 simulation.  Both the Tagged O3 simulation and the fullchem simulation used the same global O3 initial conditions.

NOTE: I am not sure to what extent the tagged O3 should match the fullchem O3, given that we are using a daily-averaged P(O3) and L(O3).  It will always probably differ in certain regions.  

**Surface**
tago3_fullchem_sfc

**500 hPa**
tago3_fullchem_500hpa

Integration tests are pending.

### CODE REVIEW

File: GeosCore/tagged_o3_mod.F90
Comment (msulprizio) [inline GeosCore/tagged_o3_mod.F90:367]: The comments here (and unit conversions below) state that prod and loss of O3 from HEMCO are in units molec/cm3/s. However, when I do ncdump on the files indicated in HEMCO_Config.rc:

```
* O3_PROD        $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.ProdLoss.$YYYY$MM$DD_0000z.nc4         Prod_Ox           2010-2019/1-12/1/0 C xyz 1        * - 1 1
* O3_LOSS        $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.ProdLoss.$YYYY$MM$DD_0000z.nc4         Loss_Ox           2010-2019/1-12/1/0 C xyz 1        * - 1 1
```

it seems the units are in fact in kg/s:

```
        float Loss_Ox(time, lev, lat, lon) ;
                Loss_Ox:long_name = "Chemical loss of Ox" ;
                Loss_Ox:units = "kg s-1" ;
                Loss_Ox:averaging_method = "time-averaged" ;

        float Prod_Ox(time, lev, lat, lon) ;
                Prod_Ox:long_name = "Chemical production of Ox" ;
                Prod_Ox:units = "kg s-1" ;
                Prod_Ox:averaging_method = "time-averaged" ;
```

Also, does it matter that technically these are prod/loss rates of Ox (as defined in fullchem.kpp) and not O3 that we are reading in via HEMCO?

**Review by msulprizio — CHANGES_REQUESTED** [review]

File: GeosCore/tagged_o3_mod.F90
Comment (yantosca) [reply GeosCore/tagged_o3_mod.F90]: Thanks @msulprizio, nice catch!.  I've removed this in commit 39b147d1fbe43179a20e44b84d89e54ac06440fd.

File: GeosCore/tagged_o3_mod.F90
Comment (yantosca) [reply GeosCore/tagged_o3_mod.F90]: I think it doesn't strictly matter that we use Prod_Ox and Loss_Ox, as these are the families that are saved out by KPP.  Ox is mostly O3 anyway.  We would have to rebuild the fullchem mechanism to add P(O3) and L(O3).

Also, I wonder if something changed between 13.0.0 and 13.4.0.  In my testing, I used 1-day output from a fullchem run and the Prod_Ox and Loss_Ox are definitely in molec/cm3/s:

```console
        float Loss_Ox(time, lev, lat, lon) ;
                Loss_Ox:long_name = "Chemical loss of Ox" ;
                Loss_Ox:units = "molec cm-3 s-1" ;
                Loss_Ox:averaging_method = "time-averaged" ;
                Loss_Ox:_Storage = "chunked" ;
                Loss_Ox:_ChunkSizes = 1, 1, 46, 72 ;
                Loss_Ox:_DeflateLevel = 1 ;
                Loss_Ox:_Shuffle = "true" ;
                Loss_Ox:_Endianness = "little" ;
                Loss_Ox:_NoFill = "true" ;
         float Prod_Ox(time, lev, lat, lon) ;
                Prod_Ox:long_name = "Chemical production of Ox" ;
                Prod_Ox:units = "molec cm-3 s-1" ;
                Prod_Ox:averaging_method = "time-averaged" ;
                Prod_Ox:_Storage = "chunked" ;
                Prod_Ox:_ChunkSizes = 1, 1, 46, 72 ;
                Prod_Ox:_DeflateLevel = 1 ;
                Prod_Ox:_Shuffle = "true" ;
                Prod_Ox:_Endianness = "little" ;
                Prod_Ox:_NoFill = "true" ;
```
I'm not sure if `Hco_EvalFld` can let you check the units of the data.  Maybe we should only convert units if the data is in molec/cm3/s.

File: GeosCore/tagged_o3_mod.F90
Comment (yantosca) [reply GeosCore/tagged_o3_mod.F90:30]: I can try to implement that.  Stay tuned.

File: GeosCore/tagged_o3_mod.F90
Comment (yantosca) [reply GeosCore/tagged_o3_mod.F90]: I just found this issue #965, which fixed a problem in the ProdLoss units.  The data in the 13.0.0 GCClassic output is in molec/cm3/s but the units are kg/s.  So I think that's OK.

**Review by msulprizio — APPROVED** [review]
My questions and suggestions for this PR have been addressed.

**Comment by yantosca** [comment]
All GEOS-Chem Classic integration tests passed:
```console
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 48
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

**Comment by map06** [comment]
@yantosca @msulprizio Is it all right to use these code changes that has only been merged to dev, or is it better to wait till all these updates are made to the main branch? I am asking this because usually, code in dev branch are  under testing phase.

**Comment by yantosca** [comment]
@map06, we usually recommend waiting until an official release.  As you say, there are several things still being added to dev as regards the fullchem simulation (from which you'll need to generate POx and LOx fields).

**Comment by map06** [comment]
@yantosca So when is the expected time period of its official release?

**Comment by yantosca** [comment]
@map06, we are working out some issues with the full-chemistry simulation that we discovered during the benchmark.  I would say hopefully sometime in the next few weeks, pending benchmark approval.

**Comment by map06** [comment]
@yantosca Ok. Thank you for the information.

**Comment by xpye98** [comment]
Hi, @yantosca, I tested the code recently. 
The concentration of O3 and other tracers is fine, but the stratospheric O3 at surface is too small, even if I ran the tagO3 simulation to spin-up for several years. 
Do you have any idea that where may go wrong? Thanks a lot!

**Comment by map06** [comment]
@yantosca Is it ok to use the dev version for tagged simulation? The official release might still take some time, and I need to run the model as soon as possible. The new version also seems to be  98 percent complete, so maybe it might not have issues running tagged simulation?

**Comment by yantosca** [comment]
@map06, you can use the dev version (reference it as 13.4.0-rc.4).

**Comment by map06** [comment]
@yantosca Thank you for the reply.

**Comment by map06** [comment]
@yantosca I tried the version `13.4.0` and  like @xpye98 mentioned, the values of stratosphere  mixing ratio in ppbv is very low. 
image. 

Is there something we still need to take care of before conducting simulation or are there still some issues?

**Comment by xpye98** [comment]
@map06 No, I haven't. My email is xpye98@outlook.com, we can discuss it together.

**Pull request merged**
